plugins/statistics.koplugin/main.lua

```diff
@@ -761,6 +767,17 @@ function ReaderStatistics:insertDB(updated_pagecount)
     end
     local id_book = self.id_curr_book
     local now_ts = os.time()
+
+    -- The current page stat, having yet no duration, will be ignored
+    -- in the insertion, and its start ts would be lost. We'll give it
+    -- to resetVolatileStats() so it can restore it
+    local cur_page_start_ts = now_ts
+    local cur_page_data = self.page_stat[self.curr_page]
+    local cur_page_data_tuple = cur_page_data and cur_page_data[#cur_page_data]
+    if cur_page_data_tuple and cur_page_data_tuple[2] == 0 then -- should always be true
+        cur_page_start_ts = cur_page_data_tuple[1]
+    end
```

Comment:
Does this actually make sense?

It feels like it's pulling the start ts from whatever was at the current page number in the *old* page layout?

Reply:
It should (I hope) obviously make sense in a context other than the onUpdateToc() page count change you reviewed the minute before :)

And in the context of onUpdateToc(), I think it still does:
- "old/new" page layout doesn't matter at this point: self.page_stat just hold tuples for a page number.
- in onUpdateToc() (your code, you thought about it), you call `self:insertDB(new_pagecount)` while `self.data.pages ` is still the old page count - you update self.data.pages = new_pagecount after insertDB.
- insertDB() just below uses (the old) self.data.pages for the pagecount associated to the page stats inserted

But errr, ok, you're talking about here backuping/restoring the latest stats for self.curr_page which is the new current page (if I trust my grammatically critical comment :)
Thinking...
I think it still doesn't matter: we started reading page 153 in the new pagecount, its latest tuple is (a_few_ms_ago, 0), the pagecount, old or new, doesn't matter. We'll be restoring and addin 153 = { (a_few_ms_ago, 0) } which is just right.

We were reading page 127/200 with latest 127 tuple being `(now-35s, 0)`
re-rendering... going to page 153/254
OnPageUpdate(153) : we update last tuple of 127: `(now-35s, 35)`. We add to 153 tuple `(now, 0)`. It sets self.curr_page=153 (but we know yet nothing about pagecount being 254)
OnUpdateToc() => insertDB() (with still self.data.pages=200, old page count)
- it inserts in DB `(127, now-35s, 35, 200)`
- it doenst insert in DB bad `(153, now, 0, 200)` because 0 duration
- ir restores to page_stats for 153: `(now, 0)`
- OnUpdateToc then sets pagecount = 254.

Later, it will be inserted in DB as `(153, now-5mn, 42, 254)`

Reply:
Oh, hey, I missed that this was an entirely different function than `onUpdateToc`, lol.

Reply:
And my post above was actually what my gramatically dubious comment was trying to explain, in a shorter way :)
May be I should add my example as a comment, because it was indeed really hard to grasp (then 2 days ago, and now :/)

Reply:
Okay, so, better question: what does this fix that simply calling resetVolatileStats w/ now_ts doesn't already?

Reply:
It fixes what my OP says it fixes: https://github.com/koreader/koreader/pull/6778#issuecomment-982779673 :)
- go to page 127, read it for 80s
- onSuspend() or showSomeStatisticsPage() or showCalendar() => insertDB() without page 127 (as duration=0) => resetVolatileStat() adds tto page 127 (now, 0)
- close stats/calendar after 5s (or onResume, that did resetVolatileStat() to (now, 0) too)
- stay on page 127 for 4s,, and turn page and be on page 128: page 127 gets (now-9s, 9) or  (now-4s, 4) and will be ignored by a future insertDB().

Later, you get a hole for page 127 that looks like it has never been read.

Use Bookmap or PageBrowser a lot (as they call a statistics method that does :insertDB()), and it really looks like you didn't read that book and spent too much time on BookMap :)

Reply:
Okay, that's much clearer, and makes perfect sense, thanks ;).